### PR TITLE
Advanced & KTX codelab fixes

### PR DESCRIPTION
**Pull request by wojtek-kalicinski**
Small fixes based on Mike's feedback and to match updated codelab text

### CODE REVIEW

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt [hunk 1]
```diff
@@ -26,13 +26,7 @@ import kotlinx.coroutines.ExperimentalCoroutinesApi
...
 import kotlinx.coroutines.Job
 import kotlinx.coroutines.channels.ConflatedBroadcastChannel
-import kotlinx.coroutines.flow.asFlow
-import kotlinx.coroutines.flow.catch
-import kotlinx.coroutines.flow.flatMapLatest
-import kotlinx.coroutines.flow.launchIn
-import kotlinx.coroutines.flow.mapLatest
-import kotlinx.coroutines.flow.onCompletion
-import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.*
```
Comment (manuelvicnt) [inline advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:29]: I'd rather use single imports in the codelabs. As sometimes there're some problems with what method to import, it's good for people to see what's the actual path for the class/method

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt [hunk 2]
```diff
@@ -127,10 +119,10 @@ class PlantListViewModel internal constructor(
      */
     fun setGrowZoneNumber(num: Int) {
         growZone.value = GrowZone(num)
-        growZoneChannel.offer(GrowZone(num))
+        growZoneFlow.value = GrowZone(num)
 
         // initial code version, remove during flow rewrite
-        launchDataLoad { plantRepository.tryUpdateRecentPlantsForGrowZoneCache(GrowZone(num)) }
+        // launchDataLoad { plantRepository.tryUpdateRecentPlantsForGrowZoneCache(GrowZone(num)) }
```
Comment (manuelvicnt) [inline advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:125]: is this needed? Revert if so

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt [hunk 3]
```diff
@@ -105,16 +98,15 @@ class PlantListViewModel internal constructor(
...
+        growZoneFlow.mapLatest { growZone ->
                 _spinner.value = true
                 if (growZone == NoGrowZone) {
                     plantRepository.tryUpdateRecentPlantsCache()
                 } else {
                     plantRepository.tryUpdateRecentPlantsForGrowZoneCache(growZone)
                 }
             }
-            .onCompletion {  _spinner.value = false }
+            .onEach { _spinner.value = false }
```
Comment (manuelvicnt) [inline advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:109]: Is this change intentional? Why did we have `onCompletion` before?

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt [hunk 4]
```diff
@@ -141,10 +133,10 @@ class PlantListViewModel internal constructor(
      */
     fun clearGrowZoneNumber() {
         growZone.value = NoGrowZone
-        growZoneChannel.offer(NoGrowZone)
+        growZoneFlow.value = NoGrowZone
 
         // initial code version, remove during flow rewrite
-        launchDataLoad { plantRepository.tryUpdateRecentPlantsCache() }
+        // launchDataLoad { plantRepository.tryUpdateRecentPlantsCache() }
```
Comment (manuelvicnt) [inline advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:139]: same here. this is the `finished_code` module, either this should be uncommented or removed

File: ktx-library-codelab/build.gradle [hunk 5]
```diff
@@ -15,7 +15,34 @@
  */
 
 buildscript {
-    ext.kotlin_version = '1.3.60'
+    ext.kotlin_version = '1.4.21'
+    ext {
+        def appcompat_ver = "1.2.0"
```
Comment (manuelvicnt) [inline ktx-library-codelab/build.gradle:20]:
```suggestion
        def appcompat_version = "1.2.0"
```

Replaced these for `_version`? It's weird to see `kotlin_version` and then other things here

File: ktx-library-codelab/step-05/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt [hunk 6]
```diff
@@ -23,11 +23,7 @@ import android.os.Looper
 import android.util.Log
 import androidx.appcompat.app.AppCompatActivity
 import androidx.lifecycle.lifecycleScope
-import com.example.android.myktxlibrary.awaitLastLocation
-import com.example.android.myktxlibrary.createLocationRequest
-import com.example.android.myktxlibrary.findAndSetText
-import com.example.android.myktxlibrary.hasPermission
-import com.example.android.myktxlibrary.showLocation
+import com.example.android.myktxlibrary.*
```
Comment (manuelvicnt) [inline ktx-library-codelab/step-05/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt:26]: Same comment about single imports

File: ktx-library-codelab/step-06/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt [hunk 7]
```diff
@@ -24,18 +24,11 @@ import androidx.appcompat.app.AppCompatActivity
 import androidx.lifecycle.Observer
 import androidx.lifecycle.asLiveData
 import androidx.lifecycle.lifecycleScope
-import com.example.android.myktxlibrary.awaitLastLocation
-import com.example.android.myktxlibrary.findAndSetText
-import com.example.android.myktxlibrary.hasPermission
-import com.example.android.myktxlibrary.locationFlow
-import com.example.android.myktxlibrary.showLocation
+import com.example.android.myktxlibrary.*
```
Comment (manuelvicnt) [inline ktx-library-codelab/step-06/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt:27]: Single imports

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt (hunk 1 above)
Comment (wojtek-kalicinski) [reply advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:29]: Argh, every time... *shakes fist at Android Studio*
Done

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt (hunk 3 above)
Comment (wojtek-kalicinski) [reply advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:109]: onCompletion was probably a mistake. Since the spinner is set to true on every emit, it should also be set to false each time after loading is complete.

File: ktx-library-codelab/step-05/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt (hunk 6 above)
Comment (wojtek-kalicinski) [reply ktx-library-codelab/step-05/app/src/main/java/com/example/android/kotlinktxworkshop/MainActivity.kt:26]: Done.

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt [hunk 8]
```diff
@@ -127,10 +123,7 @@ class PlantListViewModel internal constructor(
      */
     fun setGrowZoneNumber(num: Int) {
         growZone.value = GrowZone(num)
-        growZoneChannel.offer(GrowZone(num))
-
-        // initial code version, remove during flow rewrite
-        launchDataLoad { plantRepository.tryUpdateRecentPlantsForGrowZoneCache(GrowZone(num)) }
+        growZoneFlow.value = GrowZone(num)
```
Comment (manuelvicnt) [inline advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:126]: Nit: it's a bit confusing having `growZone` as LiveData and `growZoneFlow` as StateFlow, and then calling `.value` to both of them one after the other. Not sure what's the difference at first glance to be honest. But if it makes sense in the narrative of the codelab, leave it as is :)

**Review by manuelvicnt — APPROVED** [review]
One small nit but LGTM otherwise :) thanks for making the changes!

File: advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt (hunk 8 above)
Comment (wojtek-kalicinski) [reply advanced-coroutines-codelab/finished_code/src/main/java/com/example/android/advancedcoroutines/PlantListViewModel.kt:126]: Yeah, this is an artifact of having both LiveData and Flow in the same codelab.
I'll fix it in a next set of commits and codelab text changes
